Menus/ScenarioGUI.cpp

```diff
@@ -26,2022 +12,1051 @@
 		}
-    }
-
-    // Reset all buttons and positions of things if a new activity has been selected
-    if (newActivity)
-    {
-        m_aScenarioButton[STARTGAME]->SetVisible(false);
-        m_pStartErrorLabel->SetVisible(true);
-        m_pCPULockLabel->SetVisible(m_LockedCPUTeam != Activity::NoTeam);
-    }
-}
-
-
-//////////////////////////////////////////////////////////////////////////////////////////
-// Method:          PlayerCount
-//////////////////////////////////////////////////////////////////////////////////////////
-// Description:     Counts how many players are currently assigned to play this Activity.
-
-int ScenarioGUI::PlayerCount()
-{
-    int count = 0;
-    // Go through all the on-team non-CPU cells and see how many players are already assigned.
-    for (int player = Players::PlayerOne; player < Players::MaxPlayerCount; ++player)
-    {
-        for (int team = Activity::TeamOne; team < Activity::MaxTeamCount; ++team)
-        {
-            if (m_aapPlayerBoxes[player][team]->GetDrawType() == GUICollectionBox::Image)
-                ++count;
-        }
-    }
-    return count;
-}
-
-
-//////////////////////////////////////////////////////////////////////////////////////////
-// Method:          StartGame
-//////////////////////////////////////////////////////////////////////////////////////////
-// Description:     Sets up and starts the currently selected Activity and settings.
-// Arguments:       None.
-// Return value:    None.
-
-bool ScenarioGUI::StartGame()
-{
-    // Get the currently selected Activity
-    const Activity *pActivityPreset = m_pActivitySelect->GetSelectedItem() ? dynamic_cast<const Activity *>(m_pActivitySelect->GetSelectedItem()->m_pEntity) : 0;
-
-    if (!pActivityPreset || !m_pSelectedScene)
-        return false;
-
-    // Create the actual instance of hte Activity we want to start
-    Activity *pActivity = dynamic_cast<Activity *>(pActivityPreset->Clone());
-    GameActivity *pGameActivity = dynamic_cast<GameActivity *>(pActivity);
-
-    // Set up the basic settings
-    if (pGameActivity)
-    {
-        pGameActivity->SetDifficulty(m_pDifficultySlider->GetValue());
-
-		pGameActivity->SetFogOfWarEnabled(m_pFogOfWarCheckbox->GetCheck());
-		pGameActivity->SetRequireClearPathToOrbit(m_pRequireClearPathToOrbitCheckbox->GetCheck());
-
-
-		// If gold slider is at it max value then the amount is 'infinite' and we must set some rediculously high value
-		if (m_pGoldSlider->GetValue() == m_pGoldSlider->GetMaximum())
-			pGameActivity->SetStartingGold(1000000);
-		else
-		{
-			int startGold = m_pGoldSlider->GetValue();
-			startGold = startGold - startGold % 500;
-			pGameActivity->SetStartingGold(startGold);
+
+		if (gameActivity) {
+			int maxHumanPlayers = gameActivity->GetMaxPlayerSupport();
+			int minTeamsRequired = gameActivity->GetMinTeamsRequired();
+			if (humansInTeams > maxHumanPlayers) {
```

Comment:
This can be seriously compacted. So much duplicated code!
```cpp
if (gameActivity) {
	int maxHumanPlayers = gameActivity->GetMaxPlayerSupport();
	int minTeamsRequired = gameActivity->GetMinTeamsRequired();
	std::string errorString;

	if (humansInTeams > maxHumanPlayers) {
		errorString = "Too many players assigned! Max for this activity is " + std::to_string(maxHumanPlayers);
	} else if (minTeamsRequired > teamsWithPlayers) {
		errorString = "Assign players to at\nleast " + std::to_string(minTeamsRequired) + " of the teams!";
	} else if (teamWithHumans == 0) {
		errorString = "Assign human players\nto at least one team!";
	}

	if (errorString.empty()) {
		m_StartErrorLabel->SetText(errorString);
		m_StartErrorLabel->SetVisible(true);
	}
	m_ScenarioButtons.at(ScenarioButtons::StartGameButton)->SetVisible(errorString.empty());
}
```